client/allocrunner/taskrunner/task_runner.go

```diff
@@ -505,12 +514,14 @@ func (tr *TaskRunner) Run() {
 	var result *drivers.ExitResult
 
 	tr.stateLock.RLock()
-	dead := tr.state.State == structs.TaskStateDead
+	restoredDead := tr.state.State == structs.TaskStateDead
 	tr.stateLock.RUnlock()
 
-	// if restoring a dead task, ensure that task is cleared and all post hooks
-	// are called without additional state updates
-	if dead {
+	// If restoring a dead task, ensure that task is cleared and all post hooks
+	// are called without additional state updates.
+	// If the alloc is not terminal we must proceed until the ALLOC_RESTART
+	// loop to allow the task to run again in case the alloc is restarted.
+	if restoredDead && tr.Alloc().TerminalStatus() {
```

Comment:
This includes both the server intent and the current client state, which means we're relying on one of them to get the updated allocation. I think if we hit this for a task before the client status is updated we'll end up falling past the driver handle cleanup, which might explain the results you're seeing with the mock driver.

Reply:
Nice find. I'm not 100% if this was the problem with the tests, but it made me realize that important information here is actually whether the `Run()` method exited or not. Using the local client state fixed the problems with the `mock` driver 🎉